torch/_higher_order_ops/triton_kernel_wrap.py

```diff
@@ -61,88 +65,140 @@ def reset_table(self) -> None:
-                self.read_depth += 1
-                self.generic_visit(node)
-                self.read_depth -= 1
-                return
-        self.generic_visit(node)
-
-
-def filter_non_mutated(kernel, tensors):
+    def __add__(self, o):
+        if isinstance(o, Proxy):
+            raise Exception("Proxy operation with Proxy")
+        return self
+
+
+@dataclasses.dataclass
+class Scalar:
+    @staticmethod
+    def check(o):
+        if isinstance(o, Proxy):
+            raise Exception("Proxy operation with Scalar")
+
+    def __mul__(self, o):
+        Scalar.check(o)
+        return self
+
+    def __rmul__(self, o):
+        Scalar.check(o)
+        return self
+
+    def __add__(self, o):
+        Scalar.check(o)
+        return self
+
+    def __radd__(self, o):
+        if isinstance(o, Proxy):
+            return o
+        return self
+
+    def __gt__(self, o):
+        Scalar.check(o)
+        return self
+
+    def __lt__(self, o):
+        Scalar.check(o)
+        return self
+
+    def __and__(self, o):
+        Scalar.check(o)
+        return self
+
+    def __or__(self, o):
+        Scalar.check(o)
+        return self
+
+    def __getitem__(self, idx):
+        Scalar.check(idx)
+        return self
+
+
+# Given a triton kernel and the arguments for this kernel, this function traces
+# through the triton kernel and identifies which input pointers are mutated.
+# Tracing is done by replacing the input pointers with Proxy objects that
+# track mutation. Each triton language function is monkey patched to
+# either detect the mutation or return a fresh scalar object.
+# This tracing takes advantage of store operation being the only operation
+# that can mutate a tensor pointer.
+def identify_mutated_tensors(kernel, kwargs):
+    from triton import language as tl
     from triton.runtime.autotuner import Autotuner
 
+    # Replace tensor pointers with Proxy object. Do not replace the other
+    # inputs since tl.constexprs can be inspected by the kernel.
+    proxy_kwargs = {
+        key: (Proxy() if isinstance(value, Tensor) else value)
+        for key, value in kwargs.items()
+    }
+
     if isinstance(kernel, Autotuner):
+        if len(kernel.configs) > 0:
+            # If we are autotuning, then it doesn't matter which version gets
+            # picked for tracing purposes, so lets pick the first one
+            proxy_kwargs = {**proxy_kwargs, **kernel.configs[0].kwargs}
         kernel = kernel.fn
 
-    infos = {name: MutationInfo() for name in tensors}
-    tracker = MutationTracker(infos)
-    tracker.visit(kernel.parse())
-    return [
-        name for name, info in infos.items() if info.mutated or info.used_in_unknown
-    ]
+    default_impls: Dict[str, FunctionType] = {}
+    try:
+        # Monkey patch all triton language functions
+        for name, impl in inspect.getmembers(tl, inspect.isfunction):
+            default_impls[name] = impl
+
+            if name == "store":
```

Comment:
This approach relies on stores being the only function that can mutate a tensor pointer. I wonder how we can make this triton pin upgrade catches problems here

Reply:
I think there are a handful of others like atomic_add.

Reply:
Good catch. I thought this op was taking a triton tensor and not a pointer but I was wrong. I'll mark all atomic ops as mutational.
https://triton-lang.org/main/python-api/triton.language.html#atomic-ops

Reply:
Potentially inline assembly as well ?

Reply:
@eellison I already handled that on line 151